### PR TITLE
Use optional for compress override implementation

### LINKED ISSUE #2368: Use optional type for Reference compress
It would be cleaner to use a `bool?` here.

_Originally posted by @bernardnormier in https://github.com/zeroc-ice/ice/pull/2366#discussion_r1657837589_

### PR DESCRIPTION
**Pull request by pepone**
Fix #2368

### CODE REVIEW

File: csharp/src/Ice/Internal/Reference.cs [hunk 1]
```diff
@@ -420,9 +404,7 @@ public virtual bool Equals(Reference other)
     private Ice.ProtocolVersion _protocol;
     private Ice.EncodingVersion _encoding;
     private int _invocationTimeout;
-
-    protected bool overrideCompress_;
-    protected bool compress_; // Only used if _overrideCompress == true
+    protected bool? compress_; // Only used if _overrideCompress == true
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/Reference.cs:407]: Remove comment, make it private, rename _compress.

**Review by bernardnormier — COMMENTED** [review]
We can simplify further.

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/OutgoingConnectionFactory.java [hunk 2]
```diff
@@ -24,16 +24,6 @@ public void putOne(K key, V value) {
...
-    public boolean removeElementWithValue(K key, V value) {
-      java.util.List<V> list = this.get(key);
-      assert (list != null);
-      boolean v = list.remove(value);
-      if (list.isEmpty()) {
-        this.remove(key);
-      }
-      return v;
-    }
-
```
Comment (pepone) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/OutgoingConnectionFactory.java:36]: Unused see #2325

**Review by bernardnormier — APPROVED** [review]
Looks good, much simpler.

**Pull request merged**